File: src/main/java/co/q64/stars/tile/SeedTile.java
package co.q64.stars.tile;

import co.q64.stars.block.FormingBlock;
import co.q64.stars.block.HardBlock;
import co.q64.stars.block.RedFormedBlock;
import co.q64.stars.block.RedFormedBlock.RedFormedBlockHard;
import co.q64.stars.qualifier.SoundQualifiers.Ticking;
import co.q64.stars.type.FormingBlockType;
import co.q64.stars.type.FormingBlockTypes;
import co.q64.stars.type.forming.RedFormingBlockType;
import co.q64.stars.util.Sounds;
import lombok.Getter;
import lombok.Setter;
import net.minecraft.nbt.CompoundNBT;
import net.minecraft.tileentity.ITickableTileEntity;
import net.minecraft.tileentity.TileEntityType;
import net.minecraft.util.Direction;
import net.minecraft.util.SoundEvent;
import net.minecraft.util.math.BlockPos;
import net.minecraft.world.server.ServerWorld;

import javax.inject.Inject;

public class SeedTile extends SyncTileEntity implements ITickableTileEntity {
    protected @Inject FormingBlockTypes types;
    protected @Inject FormingBlock formingBlock;
    protected @Inject RedFormingBlockType redFormingBlockType;
    protected @Inject RedFormedBlock redFormedBlock;
    protected @Inject RedFormedBlockHard redFormedBlockHard;
    protected @Inject Sounds sounds;
    protected @Inject @Ticking SoundEvent tickingSound;

    private Direction direction;
    private @Getter @Setter int growTicks = getInitialGrowTicks();
    private @Getter @Setter boolean calculated;
    private @Getter @Setter FormingBlockType formingBlockType;
    private @Getter @Setter FormingBlockType seedType;
    private @Getter @Setter boolean primed = false;
    private @Getter @Setter boolean fruit = false;
    private @Getter @Setter double multiplier = 1;

    @Inject
    protected SeedTile(TileEntityType<? extends SeedTile> type) {
        super(type);
    }

    @Inject
    protected void setupDefault(FormingBlockTypes types) {
        this.formingBlockType = types.getDefault();
        this.seedType = types.getDefault();
    }

    public void read(CompoundNBT compound) {
        super.read(compound);
        formingBlockType = types.get(compound.getInt("formingBlockType"));
        seedType = types.get(compound.getInt("seedType"));
        calculated = compound.getBoolean("calculated");
        primed = compound.getBoolean("primed");
        growTicks = compound.getInt("growTicks");
        fruit = compound.getBoolean("fruit");
    }

    public CompoundNBT write(CompoundNBT compound) {
        CompoundNBT result = super.write(compound);
        result.putInt("formingBlockType", formingBlockType.getId());
        result.putInt("seedType", seedType.getId());
        result.putBoolean("calculated", calculated);
        result.putBoolean("primed", primed);
        result.putInt("growTicks", growTicks);
        result.putBoolean("fruit", fruit);
        return result;
    }

    protected boolean hasSeed() {
        return true;
    }

    protected int getInitialGrowTicks() {
        return 20;
    }

    public void tick() {
        if (!world.isRemote) {
            if (!hasSeed()) {
                return;
            }
            if (growTicks == getInitialGrowTicks()) {
                if (isPrimed()) {
                    sounds.playSound((ServerWorld) world, pos, tickingSound, 1f);
                }
            }
            if (growTicks == 0) {
                if (isPrimed()) {
                    redFormingBlockType.explode((ServerWorld) world, pos, false);
                } else {
                    direction = seedType.getInitialDirection(world, pos);
                    if (direction == null) {
                        return;
                    }
                    BlockPos target = pos.offset(direction);
                    world.setBlockState(target, formingBlock.getDefaultState());

                    FormingTile tile = (FormingTile) world.getTileEntity(target);
                    if (tile == null) {
                        System.out.println("null tile on seed block grow " + target.toString());
                    } else {
                        tile.setFirst(true);
                        tile.setDirection(direction);
                        tile.setMultiplier(multiplier);
                        tile.setup(seedType);
                        tile.setHard(getBlockState().getBlock() instanceof HardBlock);
                        tile.setCalculated(true);
                    }
                }
            }
            if (growTicks < 0) {
                boolean hard = getBlockState().getBlock() instanceof HardBlock;
                if (direction == null) {
                    return;
                }
                if (world.getBlockState(pos.offset(direction)).getBlock() != formingBlock) {
                    if (formingBlockType instanceof RedFormingBlockType) {
                        world.setBlockState(pos, hard ? redFormedBlockHard.getDefaultState() : redFormedBlock.getDefaultState());
                    } else {
                        world.setBlockState(pos, hard ? formingBlockType.getFormedBlockHard().getDefaultState() : formingBlockType.getFormedBlock().getDefaultState());
                    }
                }
            }
            growTicks--;
        }
    }
}

File: src/main/kotlin/co/q64/stars/tile/SeedTile.kt
package co.q64.stars.tile

import co.q64.stars.block.FormingBlock
import co.q64.stars.block.HardBlock
import co.q64.stars.block.RedFormedBlock
import co.q64.stars.block.RedFormedBlockHard
import co.q64.stars.type.BlueFormingBlockType
import co.q64.stars.type.FormingBlockType
import co.q64.stars.type.RedFormingBlockType
import net.minecraft.nbt.CompoundNBT
import net.minecraft.tileentity.ITickableTileEntity
import net.minecraft.tileentity.TileEntityType
import net.minecraft.util.Direction
import net.minecraft.world.World
import net.minecraft.world.server.ServerWorld

open class SeedTile(type: TileEntityType<*> = seedTileType) : SyncTileEntity(type), ITickableTileEntity {
    var direction: Direction? = null
    var ticks = 20
    var active = false
    var ready = false
    var type: FormingBlockType = BlueFormingBlockType
    var seed: FormingBlockType = BlueFormingBlockType
    var primed = false
    var fruit = false
    var multiplier = 1.0

    override fun read(compound: CompoundNBT) = with(compound) {
        super.read(this)
        type = FormingBlockType.fromId(getString("type"))
        seed = FormingBlockType.fromId(getString("seed"))
        ready = getBoolean("ready")
        primed = getBoolean("primed")
        fruit = getBoolean("fruit")
        ticks = getInt("ticks")
        active = getBoolean("active")
    }

    override fun write(compound: CompoundNBT): CompoundNBT = with(compound) {
        putString("type", type.id)
        putString("seed", seed.id)
        putBoolean("ready", ready)
        putBoolean("primed", primed)
        putBoolean("fruit", fruit)
        putBoolean("active", active)
        return super.write(compound)
    }


    override fun tick() {
        world?.let { world ->
            if (world.isRemote && !active) return
            if (ticks == 0) {
                (world as? ServerWorld)?.getClosestPlayer(pos.x.toDouble(), pos.z.toDouble(), 1000.0)?.let {
                    // TODO level type
                }
            }
            var count = 0
            if (ticks == 0) {
                grow(world)
            } else if (ticks < 0) {
                check(world)
            }
            ticks--
        }
    }

    private fun grow(world: World) {
        if (primed) {
            RedFormingBlockType.explode(world as ServerWorld, pos, false)
            return
        }
        direction = seed.firstDirection(world, pos)
        direction?.let {
            val target = pos.offset(it)
            world.setBlockState(target, FormingBlock.defaultState)
            (world.getTileEntity(target) as? FormingTile)?.apply {
                first = true
                direction = it
                multiplier = this@SeedTile.multiplier
                hard = blockState.block is HardBlock
                type = seed
                ready = true
            }
        }
    }

    private fun check(world: World) {
        direction?.let {
            val hard = blockState.block is HardBlock
            if (world.getBlockState(pos.offset(it)).block != FormingBlock) {
                if (type is RedFormingBlockType) {
                    world.setBlockState(pos, if (hard) RedFormedBlockHard.defaultState else RedFormedBlock.defaultState)
                } else {
                    world.setBlockState(pos, if (hard) type.formedHard.defaultState else type.formed.defaultState)
                }
            }
        }
    }
}